src/protocol/attribution/accumulate_credit.rs

```diff
@@ -0,0 +1,300 @@
+        ctx: ProtocolContext<'_, N>,
+    ) -> Result<Batch<AccumulateCreditOutputRow<F>>, BoxError> {
+        #[allow(clippy::cast_possible_truncation)]
+        let num_rows = self.input.len() as RecordIndex;
+
+        // 1. Create credit and stop_bit vectors
+        // These vectors are updated in each iteration to help accumulate values and determine when to stop accumulating.
+
+        let one = Replicated::one(ctx.mesh().identity());
+        let mut stop_bits: Batch<Replicated<F>> = vec![one; num_rows as usize].into();
+
+        let mut credits: Batch<Replicated<F>> = self
+            .input
+            .iter()
+            .map(|x| x.value)
+            .collect::<Vec<_>>()
+            .into();
+
+        // 2. Accumulate (up to 4 multiplications)
+        //
+        // For each iteration (`log2(input.len())`), we access each node in the
+        // list to accumulate values. The accumulation can be optimized to the
+        // following arithmetic circuit.
+        //
+        //   b = current.stop_bit * successor.helper_bit * successor.trigger_bit;
+        //   new_credit[current_index] = current.credit + b * successor.credit;
+        //   new_stop_bit[current_index] = b * successor.stop_bit;
+        //
+        // Each list element interacts with exactly one other element in each
+        // iteration, and the interaction do not depend on the calculation results
+        // of other elements, allowing the algorithm to be executed in parallel.
+
+        let mut iteration_step = IterStep::new("iteration");
+
+        // generate powers of 2 that fit into input len. If num_rows is 15, this will produce [1, 2, 4, 8]
+        for step_size in std::iter::successors(Some(1u32), |prev| prev.checked_mul(2))
+            .take_while(|&v| v < num_rows)
+        {
+            let end = num_rows - step_size;
+            let mut accumulation_futures = Vec::with_capacity(end as usize);
+
+            let ctx = ctx.narrow(iteration_step.next());
+            let mut multiply_step = IterStep::new("multiply");
+
+            // for each input row, create a future to execute secure multiplications
+            for i in 0..end {
+                let current = AccumulateCreditInputRow {
+                    stop_bit: stop_bits[i],
+                    credit: credits[i],
+                    report: self.input[i],
+                };
+                let successor = AccumulateCreditInputRow {
+                    stop_bit: stop_bits[i + step_size],
+                    credit: credits[i + step_size],
+                    report: self.input[i + step_size],
+                };
+
+                accumulation_futures.push(Self::get_accumulated_credit(
+                    ctx.narrow(multiply_step.next()),
+                    current,
+                    successor,
+                    iteration_step.is_first_iteration(),
+                ));
+            }
+
+            let results = try_join_all(accumulation_futures).await?;
```

Comment:
@martinthomson in many protocols there seem to be a pattern of generating large number of futures and then polling them all. Do you see it as a problem, given that no work is done until the first poll?